FEA add PredictionErrorDisplay (#18020)
scikit-learn/scikit-learn: doc/modules/model_evaluation.rst

  >>> d2_absolute_error_score(y_true, y_pred)
  0.0

Visual evaluation of regression models
--------------------------------------

The :class:`~sklearn.metrics.PredictionErrorDisplay` displays allows to visually

> inspect the quality of regression models. The quality of a regression model can

Reviewer: Maybe mention that it's not applicable for quantile regression as Christian pointed out

be assessed using two different plots as shown below:

.. image:: ../auto_examples/model_selection/images/sphx_glr_plot_cv_predict_001.png
   :target: ../auto_examples/model_selection/plot_cv_predict.html
   :scale: 75
   :align: center

The left-hand side plot shows the predicted values vs. the actual values. A
perfect regression model will display data point on the diagonal of this plot.
Further away from this optimal line, the larger the error of the model.

The right-hand side plot shows the residuals, i.e. the difference between the
actual values and the predicted values, vs. the predicted values. This type of
plot give an indication of the homoscedasticity of the model. A perfect
regression model will display data points uniformly distributed around the
horizontal axis. If the distribution is not uniform, then it indicates us that
there is a relationship between the value of the target and the residuals, e.g.
the error increases when the true target increases.

> Refer to the example below to see an model evaluation that make use of this

Reviewer: ```suggestion
Refer to the example below to see a model evaluation that make use of this
```

  >>> d2_absolute_error_score(y_true, y_pred)
  0.0

Visual evaluation of regression models
--------------------------------------

The :class:`~sklearn.metrics.PredictionErrorDisplay` displays allows to
visually inspect the quality of regression models. The quality of a regression
model can be assessed using two different plots as shown below:

.. image:: ../auto_examples/model_selection/images/sphx_glr_plot_cv_predict_001.png
   :target: ../auto_examples/model_selection/plot_cv_predict.html
   :scale: 75
   :align: center

The plot on the left shows the predicted values vs. the actual values. A
perfect regression model would display data point on the diagonal defined by
predicted = actual values. The further away from this optimal line, the larger
the error of the model. Note that this analysis is only meaningful when the
predicted values is the expectation. This plot cannot be used when the model
predicts a specific quantile, e.g. using
:class:`~sklearn.linear_model.QuantileRegressor`.

> The right-hand side plot shows the residuals, i.e. the difference between the
> actual values and the predicted values, vs. the predicted values. This type of
> plot give an indication of the `homoscedasticity
> <https://en.wikipedia.org/wiki/Homoscedasticity_and_heteroscedasticity>`_ of
> the model. A perfect regression model will display data points uniformly
> distributed around the horizontal axis. If the distribution is not uniform,
> then it indicates us that there is a relationship between the value of the
> target and the residuals, e.g. the error increases when the true target
> increases.

Reviewer 2: There is no reason for the "perfect model" to display homoschedastic residuals whenever the irreducible error is non-zero. If the irreducible error is null, then
the perfect model has zero residuals and the residuals homoschedastic as a consequence
but the homoschedasticity of a deterministic "random" variable is rather pointless.

Also the horizontal axis is not a density function so it does not make sense to speak about a uniform distribution when reading this plot... The distributions of residuals is rather to be "read vertically" for a given narrow, horizontal slice of `y_pred`.

Let' me suggest the following instead"
 
```suggestion
The right-hand side plot shows the residuals, i.e. the difference between the
actual values and the predicted values, vs. the predicted values.

This plot makes it easier to visualize if the residuals follow
and `homoscedastic or heteroschedastic
<https://en.wikipedia.org/wiki/Homoscedasticity_and_heteroscedasticity>`_
distribution. In particular, if the true distribution of `y|X` is Poisson
or Gamma distributed, it is expected that the variance of the residuals
of the optimal model to grow with the predicted value of `E[y|X]` (either linearly
for Poisson or quadratically for Gamma).

We can also use this plot to check if the residuals are Gaussian distributed
with a constant variance (homeschedastic residuals) which the assumption made
when fitting linear least squares regression model (see
:class:`sklearn.linear_mnodel.LinearRegression` and
:class:`sklearn.linear_mnodel.Ridge`). If this is not the case, and in particular if
the residual plot some banana-shaped structure, this is a hint that the model is
likely mis-specified and that non-linear feature engineering or switching to a
non-linear regression model might be useful.
```

Reviewer 2: For reference, here is more information on how to interpret those kinds of plots:

https://www.qualtrics.com/support/stats-iq/analyses/regression-guides/interpreting-residual-plots-improve-regression/

I think it would be great to write some kind of advanced tutorial to diagnose various kinds of model misspecification problems for linear regression (penalized or not) but this should be done in a dedicated PR.

  >>> d2_absolute_error_score(y_true, y_pred)
  0.0

Visual evaluation of regression models
--------------------------------------

The :class:`~sklearn.metrics.PredictionErrorDisplay` class allows to
visually inspect the quality of regression models. The quality of a regression
model can be assessed using two different plots as shown below:

.. image:: ../auto_examples/model_selection/images/sphx_glr_plot_cv_predict_001.png
   :target: ../auto_examples/model_selection/plot_cv_predict.html
   :scale: 75
   :align: center

The plot on the left shows the actual values vs predicted values. For a
noise-free regression task, a perfect regression model would display data points
on the diagonal defined by predicted = actual values. The further away from this
optimal line, the larger the error of the model. In a more realistic setting with
irreducible noise, that is, when not all the variations of `y` can be explained
by features in `X`, then the best model would lead to a cloud of points densely
arranged around the diagonal.

Note that the above holds when the predicted values is the expected value of `y`

> given `X`. This is typically the case for regression model that asymptotically

Reviewer 2: > If we keep them, the log loss should be added.

This diagram is only valid when compute from outputs of scikit-learn regressors with continuous outcome variables. Residual plots for binary outcome variables seem uncommon/unexpected to me, and cannot be computed at all for multiclass outcomes.

Ok to drop the "asymptotically" for the sake of simplicity.

Also:

```suggestion
given `X`. This is typically the case for regression models that asymptotically
```

--------------------------------------

Among methods to assess the quality of regression models, scikit-learn provides
the :class:`~sklearn.metrics.PredictionErrorDisplay` class. It allows to
visually inspect the prediction errors of a model in two different manners.

.. image:: ../auto_examples/model_selection/images/sphx_glr_plot_cv_predict_001.png
   :target: ../auto_examples/model_selection/plot_cv_predict.html
   :scale: 75
   :align: center

The plot on the left shows the actual values vs predicted values. For a
noise-free regression task aiming to predict the (conditional) expectation of
`y`, a perfect regression model would display data points on the diagonal
defined by predicted equal to actual values. The further away from this optimal
line, the larger the error of the model. In a more realistic setting with
irreducible noise, that is, when not all the variations of `y` can be explained
by features in `X`, then the best model would lead to a cloud of points densely
arranged around the diagonal.

Note that the above only holds when the predicted values is the expected value
of `y` given `X`. This is typically the case for regression models that

> asymptotically minimize the mean squared error objective function or more
> generally the :ref:`mean Tweedie deviance <mean_tweedie_deviance>` for any
> value of its "power" parameter.

Reviewer 2: Lets drop the "asymptotically" as discussed earlier:

```suggestion
minimize the mean squared error objective function or more generally the
:ref:`mean Tweedie deviance <mean_tweedie_deviance>` for any value of its
"power" parameter.
```

When plotting the predictions of an estimator that predicts a quantile
of `y` given `X`, e.g. :class:`~sklearn.linear_model.QuantileRegressor`

> or any other model asymptotically minimizing the :ref:`pinball loss
> <pinball_loss>`, a fraction of the points are either expected to lie above or
> below the diagonal depending on the estimated quantile level.

Reviewer 2: Same:

```suggestion
or any other model minimizing the :ref:`pinball loss <pinball_loss>`, a
fraction of the points are either expected to lie above or below the diagonal
depending on the estimated quantile level.
```

All in all, while intuitive to read, this plot does not really inform us on
what to do to obtain a better model.

The right-hand side plot shows the residuals (i.e. the difference between the
actual and the predicted values) vs. the predicted values.

This plot makes it easier to visualize if the residuals follow and
`homoscedastic or heteroschedastic
<https://en.wikipedia.org/wiki/Homoscedasticity_and_heteroscedasticity>`_
distribution.

In particular, if the true distribution of `y|X` is Poisson or Gamma
distributed, it is expected that the variance of the residuals of the optimal

> model to grow with the predicted value of `E[y|X]` (either linearly for

Reviewer 2: ```suggestion
model would grow with the predicted value of `E[y|X]` (either linearly for
```

Poisson or quadratically for Gamma).

> Since a predictive models do not adapt themselves to the particularities of the
> true distribution, it comes to us at choosing the optimal modelisation given
> the data at hand. This plot is therefore useful for making a choice between
> different family of models.
> 

Reviewer 2: I don't think the first sentence is true. Any non-trivial machine learning model
adapts its parameters to best match the empirical distribution.

But even if you refer to shape of the true conditional distribution and try to match it by choosing different fitting criterions or parametrizations, it would still be possible to design a metaestimator that runs and an internal CV and tries many estimator classes.

Anyways, I don't think this paragraph brings any useful information beyond what is explained in more specific previous and following paragraphs. Instead it can bring confusion so I would suggest to just drop it.

```suggestion
```

> We can also use this plot to check if the residuals are Gaussian distributed
> with a constant variance (homeschedastic residuals) which is the assumption
> made when fitting linear least squares regression model (see
> :class:`sklearn.linear_mnodel.LinearRegression` and
> :class:`sklearn.linear_mnodel.Ridge`). If this is not the case, and in
> particular if the residual plot some banana-shaped structure, this is a hint
> that the model is likely mis-specified and that non-linear feature engineering
> or switching to a non-linear regression model might be useful.

Reviewer 2: ```suggestion
When fitting a linear least squares regression model (see :class:`sklearn.linear_mnodel.LinearRegression` and
:class:`sklearn.linear_mnodel.Ridge`), we can use this plot to check
if some of the `model assumptions
<https://en.wikipedia.org/wiki/Ordinary_least_squares#Assumptions>`_
are met, in particular that the residuals should be uncorrelated, their
expected value should be null and that their variance should be constant
(homoschedasticity).

If this is not the case, and in particular if the residual plot some
banana-shaped structure, this is a hint that the model is likely mis-specified
and that non-linear feature engineering or switching to a non-linear regression
model might be useful.
```

Reviewer 2: Do you agree with the above suggestion @lorentzenchr?

Reviewer 2: I do not necessarily want to go into more details about estimator efficiency. I think people can go read the wikipedia article that I linked which also links to the article about the Gauss Markov Theorem on one hand and on generalized least squares on the other hand (for correlated residuals with a known correlation structure).
